pip/pip-269.md

```diff
@@ -0,0 +1,239 @@
+```java
+public static class CursorModifyingException extends ManagedLedgerException {};
+
+public static class CursorEpochConflictException extends ManagedLedgerException {};
+```
+
+<strong>ManagedCursorImpl.java</strong>(maybe instead `synchronized` to a `lock`)
+```java
+/** Indicates how many times the `read position` of the cursor has been changed by the **Occasional Read Position Change** **/
+private volatile int epoch;
+/**
+ * 0: no task is running.
+ * 1: the task is running 
+ */
+private volatile int isEpochIncreaseRunning;
+
+/** Return current epoch, will not care about if there has a running epoch increase task. **/
+public int getEpoch();
+
+/**
+* Set `isEpochIncreaseRunning` to true and return the current epoch.
+* throws CursorEpochConflictException if there a running epoch increase task is in progress. 
+*/
+public synchronized int startIncreaseEpoch(Supplier<Position> newReadPositionCalcultor);
+
+/**
+ * Compare and set `isEpochIncreaseRunning` to `0` and return the `isEpochIncreaseRunning` before the update. If there is no epoch increment in progress, will not increase and just return `0`.
+ */
+public synchronized int endIncreaseEpoch();
+
+/** Only used for the scenario Sequentially Read, return true if set succeed. **/
+public synchronized boolean setReadPosition(Position newPosition, int expectedCursorEpoch);
+
+/** Return false if there is a running epoch increase task or epoch was changed. **/
+public boolean isEpochOutdated(int expectedSpoch);
+```
+
+**CursorStats.java**
+
+```java
+public int epoch;
+public int isEpochIncreaseRunning;
+```
+
+A demo of `rewind` in Key_shared mode:
+
+```java
+try {
+  cursor.startIncreaseEpoch(() ->  markDeletedPostion + 1);
+  dispatcher.clearDelayedMessageTracker();
+  dispatcher.clearRedeliveryMessages();
+  dispatcher.clearRecentJoinedConsumers();
+  if (cursor.endIncreaseEpoch() == 2 ) {
+    readMoreEntries(); // compensate read once.
+  }
+} catch (CursorEpochConflictException ex) {
+  // do retry or something else.
+}
+```
+
+A demo of **Sequentially Read**:
+
+```java
+public synchronized void readMoreEntries() {
+  ReadEntriesCallback cb = this;
+  int entriesCountToRead = 10;
+  int epoch = cursor.getEpoch();
+  cursor.asyncRead(entriesCountToRead, cb, epoch);
+}
+
+public final synchronized void readEntriesComplete(List<Entry> entries, Object ctx) {
+  int cursorEpochWhenStartRead = (int) ctx;
+  // Normally “readEntriesComplete” and **Occasional Read Position Change** are mutually exclusive.
+  if (cursor.isEpochOutdated(cursorEpochWhenStartRead)) {
+    entries.forEach(Entry::release);
+    return;
+  }
+  // send messages to client.
+}
+```
+
+### Metrics
+
+**pulsar_cursor_epoch_increase_count**
+- Description: how many times the `read position` of the cursor has been changed by the **Occasional Read Position Change**
+- type: Counter
+- labels: `[cluster, namespace, topic, subscription]`
+- unit: times
+
+**pulsar_cursor_epoch_increase_task_state**
```

Comment:
Better count, as it will rapidly change, sampling every 30sec you won't see anything.

Reply:
There is already a counter that does this `pulsar_cursor_epoch_increase_count`, and this one just represents a state of if there are a running Read Position Reset.

Reply:
Ok. makes sense